x/examples/test-connectivity/main.go

```diff
@@ -120,7 +134,9 @@ func init() {
 }
 func newTCPTraceDialer(
 	onDNS func(ctx context.Context, domain string) func(di httptrace.DNSDoneInfo),
-	onDial func(ctx context.Context, network, addr string, connErr error)) transport.StreamDialer {
+	onDial func(ctx context.Context, network, addr string, connErr error),
```

Comment:
This design puts the onus on the user to match dial start and end. That also complicates the testing code.

Please change the design so that we have a single onDial that is called on the beginning of the dial, and returns the onDialDone handler, like we do for DNS.

In the case of TCP/UDP, you will need a map (net, addr) -> done handler that you can create here.

Reply:
@fortuna I am going to iterate on the proposed design. Ideally the user should just run the test and get report with a clean API that is part of the SDK (/x) and `test-connectivity` code would just use it.